### PR TITLE
expression, executor: rewrite builtin func degrees

### PR DESCRIPTION
**Pull request by youyouhu**

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
@youyouhu Thank you for your PR,
`newBaseBuiltinFuncWithTp()`  has been updated,
please update your code.

**Comment by youyouhu** [comment]
There is something  weird. In TestCos, the case {"sdfgsfsdf", float64(0), false, true} is right. In TestDegrees, the case {"abc", float64(0), false, true} is wrong. b.args[0].EvalReal(row, b.ctx.GetSessionVars().StmtCtx) return (0, false, nil) in both cases

File: expression/builtin_math_test.go [hunk 1]
```diff
@@ -449,10 +449,10 @@ func (s *testEvaluatorSuite) TestDegrees(c *C) {
 		{float64(1), float64(57.29577951308232), false, false},
 		{float64(math.Pi), float64(180), false, false},
 		{float64(-math.Pi / 2), float64(-90), false, false},
-		{"", float64(0), false, true},
+		{"", float64(0), false, false},
 		{"-2", float64(-114.59155902616465), false, false},
-		{"abc", float64(0), false, true},
-		{"+1abc", float64(0), false, true},
+		{"abc", float64(0), false, false},
```
Comment (XuHuaiyu) [inline expression/builtin_math_test.go]: from line 434 to line 438
set the `IgnoreTruncate` to false for this statement context,
so StrToFloat("abc") will not get a truncate error.

@youyouhu

File: expression/builtin_math.go [hunk 2]
```diff
@@ -914,35 +914,27 @@ type degreesFunctionClass struct {
...
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
 type builtinDegreesSig struct {
-	baseBuiltinFunc
+	baseRealBuiltinFunc
 }
 
-// eval evals a builtinDegreesSig.
+// evalreal evals a builtinDegreesSig.
```
Comment (XuHuaiyu) [inline expression/builtin_math.go]: s/evalreal/evalReal

File: executor/executor_test.go [hunk 3]
```diff
@@ -966,6 +966,12 @@ func (s *testSuite) TestBuiltin(c *C) {
 	result = tk.MustQuery("select sin('abcd')")
 	result.Check(testkit.Rows("0"))
 
+	//test degrees
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: add test cases in sql level in executor/executor_test.go/TestMathBuiltin

Comment (plutolove) [inline executor/executor_test.go]: ok

File: expression/builtin_math_test.go (hunk 1 above)
Comment (plutolove) [inline expression/builtin_math_test.go]: I got it

**Comment by youyouhu** [comment]
I have fixed all
@XuHuaiyu

File: expression/builtin_math_test.go [hunk 4]
```diff
@@ -465,6 +460,7 @@ func (s *testEvaluatorSuite) TestDegrees(c *C) {
 		c.Assert(tp.Flag, Equals, uint(mysql.BinaryFlag))
 		c.Assert(tp.Flen, Equals, 23)
 		d, err := f.Eval(nil)
+		fmt.Println(t.args)
```
Comment (XuHuaiyu) [inline expression/builtin_math_test.go]: remove this

File: executor/executor_test.go [hunk 5]
```diff
@@ -1095,6 +1089,21 @@ func (s *testSuite) TestBuiltin(c *C) {
...
+func (s *testSuite) TestMathBuiltin(c *C) {
+	defer func() {
+		s.cleanEnv(c)
+		testleak.AfterTest(c)()
+	}()
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+
+	//test degrees
+	result := tk.MustQuery("select degrees(0)")
```
Comment (XuHuaiyu) [inline executor/executor_test.go]: select degrees("0"), degrees("1")
testkit.Rows("result0 result1")

Comment (XuHuaiyu) [inline executor/executor_test.go]: select degrees("0"), degrees("1")
testkit.Rows("result0 result1")

Comment (XuHuaiyu) [inline executor/executor_test.go]: move cos, sin, tan in TestStringBuiltin to this func.

Comment (youyouhu) [inline executor/executor_test.go]: ok

**Comment by youyouhu** [comment]
I have fixed all
@XuHuaiyu

File: expression/builtin_math_test.go [hunk 6]
```diff
@@ -429,39 +429,46 @@ func (s *testEvaluatorSuite) TestSign(c *C) {
...
+		} else {
+			c.Assert(err, IsNil)
+			if t.isNil {
+				c.Assert(d.Kind(), Equals, types.KindNull)
+			} else {
+				c.Assert(d.GetFloat64(), Equals, t.expected)
+			}
+		}
 	}
 }
```
Comment (XuHuaiyu) [inline expression/builtin_math_test.go]: Please add test case for isDetermistic.

Comment (youyouhu) [inline expression/builtin_math_test.go]: Does cos,sin,tan need to add it？
@XuHuaiyu

Comment (XuHuaiyu) [inline expression/builtin_math_test.go]: yes

**Comment by youyouhu** [comment]
I have fixed all
@XuHuaiyu

**Comment by XuHuaiyu** [comment]
Good job, LGTM
PTAL @zz-jason

**Pull request merged**